### PR TITLE
circulation: adapt reroils

### PR DESCRIPTION
**Pull request by BadrAly**
Adapts reroils after the upgrade to invenio-circulation v1.0.0a21.

The two transitions ItemOnLoanToItemReturned and ItemOnLoanToItemReturned are
temporarily updated and inserted into reroils until a fix for the issue
https://github.com/inveniosoftware/invenio-circulation/issues/127 is available.

This PR expects the fix for the ITEM_AT_DESK problem in the current version
of invenio-circulation.

Some timezone units testing are disabled until a fix is given in a later PR.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Dependencies

This PR requires the `rero/invenio-circulation.item_at_desk` branch
## Why are you opening this PR?

https://tree.taiga.io/project/rero21-reroils/task/1431?kanban-status=1224894

## How to test?

`bootstrap + run_tests.sh`

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: rero_ils/config.py [hunk 1]
```diff
@@ -1586,18 +1588,20 @@ def _(x):
 }
 
 #: Invenio circulation configuration.
-CIRCULATION_ITEM_EXISTS = Item.get_record_by_pid
+CIRCULATION_ITEM_EXISTS = item_exists
```
Comment (zannkukai) [inline rero_ils/config.py:1591]: My point of view : I thing the best place for this method is to place it into the `modules.items.api:Item` class as a static method. Maybe other comments to confirm of not this point of view.

File: rero_ils/modules/items/api.py [hunk 2]
```diff
@@ -285,34 +287,57 @@ def holding_pid(self):
...
     def get_loans_by_item_pid(cls, item_pid):
         """Return any loan loans for item."""
-        results = current_circulation.loan_search_cls.filter(
-            'term', item_pid=item_pid).source(includes='pid').scan()
+        item_pid_object = item_pid_to_object(item_pid)
+        search_cls = current_circulation.loan_search_cls
+        search = search_cls().filter(
+            'term', item_pid__value=item_pid_object['value']).filter(
+            'term', item_pid__type=item_pid_object['type'])
+        results = search.source(includes='pid').scan()
```
Comment (zannkukai) [inline rero_ils/modules/items/api.py:332]: You can merge to skip `search` variable : 
```
results = current_circulation.loan_search_cls()
            .filter('term', item_pid__value=item_pid_object['value'])
            .filter('term', item_pid__type=item_pid_object['type'])
            .source(includes='pid')
            .scan()
```

File: rero_ils/modules/items/api.py [hunk 3]
```diff
@@ -358,13 +383,14 @@ def get_pendings_loans(cls, library_pid=None, sort_by='transaction_date'):
...
-            .filter('term', library_pid=library_pid)\
-            .sort({sort_by: {"order": order_by}})
-        results = search.scan()
+
+        search_cls = current_circulation.loan_search_cls
+        search = search_cls().params(preserve_order=True).filter(
+            'term', state='PENDING').filter(
+            'term', library_pid=library_pid).sort(
+                {sort_by: {"order": order_by}})
+        results = search.source(includes='pid').scan()
```
Comment (zannkukai) [inline rero_ils/modules/items/api.py:392]: same here (with indentation to be more human readable) :
```
results = current_circulation.loan_search_cls()
            .params(preserve_order=True)
            .filter('term', state='PENDING')
            .filter('term', library_pid=library_pid)
            .sort({sort_by: {"order": order_by}})
            .source(includes='pid')
            .scan()
```

File: rero_ils/modules/items/api.py [hunk 4]
```diff
@@ -383,11 +409,13 @@ def get_checked_out_loans(
...
-            .filter('term', state='ITEM_ON_LOAN')\
-            .filter('term', patron_pid=patron_pid)\
-            .sort({sort_by: {"order": order_by}}).scan()
+        search_cls = current_circulation.loan_search_cls
+        search = search_cls().params(preserve_order=True).filter(
+            'term', state='ITEM_ON_LOAN').filter(
+            'term', patron_pid=patron_pid).sort(
+                {sort_by: {"order": order_by}})
+        results = search.source(includes='pid').scan()
+
```
Comment (zannkukai) [inline rero_ils/modules/items/api.py:418]: same here

File: rero_ils/modules/items/api_views.py [hunk 5]
```diff
@@ -229,6 +230,7 @@ def requested_loans(library_pid):
         library_pid=library_pid, sort_by=sort_by)
     metadata = []
     for item, loan in items_loans:
+        a = loan.dumps_for_circulation()
```
Comment (zannkukai) [inline rero_ils/modules/items/api_views.py:233]: seems `a` is never more used ? if yes, remove this line.

File: rero_ils/modules/loans/api.py [hunk 6]
```diff
@@ -98,53 +99,26 @@ def create(cls, data, id_=None, delete_pid=True,
...
+                self[field] = str2datetime(self[field])
+
+    def date_fields2str(self):
+        """Convert Python datetime fields to string."""
+        for field in self.DATE_FIELDS:
+            if field in self:
+                self[field] = self[field].date().isoformat()
+        for field in self.DATETIME_FIELDS:
+            if field in self:
+                self[field] = self[field].isoformat()
```
Comment (zannkukai) [inline rero_ils/modules/loans/api.py:115]: Maybe these 2 functions best places are into `modules/utils.py` ?

File: rero_ils/modules/loans/api.py [hunk 7]
```diff
@@ -292,12 +274,12 @@ def get_request_by_item_pid_by_patron_pid(item_pid, patron_pid):
...
-        .params(preserve_order=True)\
-        .filter('term', patron_pid=patron_pid)\
-        .sort({'transaction_date': {'order': 'asc'}})\
-        .scan()
+    search_cls = current_circulation.loan_search_cls
+    search = search_cls().filter(
+        'term', patron_pid=patron_pid).params(
+        preserve_order=True).sort(
+        {'transaction_date': {'order': 'asc'}}).source(['pid'])
+    results = search.source(includes='pid').scan()
```
Comment (zannkukai) [inline rero_ils/modules/loans/api.py:282]: as previous comment

File: rero_ils/modules/loans/api.py [hunk 8]
```diff
@@ -343,13 +325,14 @@ def patron_profile_loans(patron_pid):
...
-        .sort({'transaction_date': {'order': 'desc'}})\
-        .scan()
+    search_cls = current_circulation.loan_search_cls
+    search = search_cls().filter(
+        'term', item_pid=item_pid).params(
+        preserve_order=True).exclude(
+        'terms', state=['PENDING', 'CREATED']).sort(
+        {'transaction_date': {'order': 'desc'}}).source(
+        ['pid'])
+    results = search.source(includes='pid').scan()
```
Comment (zannkukai) [inline rero_ils/modules/loans/api.py:335]: same here

File: rero_ils/modules/loans/api.py [hunk 9]
```diff
@@ -362,19 +345,20 @@ def get_due_soon_loans():
...
-        .filter('term', state='ITEM_ON_LOAN')\
-        .sort({'transaction_date': {'order': 'asc'}})\
-        .scan()
+    search_cls = current_circulation.loan_search_cls
+    search = search_cls().filter(
+        'term', state='ITEM_ON_LOAN').params(
+        preserve_order=True).sort(
+        {'transaction_date': {'order': 'asc'}}).source(
+        ['pid'])
+    results = search.source().scan()
```
Comment (zannkukai) [inline rero_ils/modules/loans/api.py:354]: same here

File: rero_ils/modules/loans/api.py [hunk 10]
```diff
@@ -385,12 +369,13 @@ def get_overdue_loans():
...
```
Comment (zannkukai) [inline rero_ils/modules/loans/api.py:378]: same here

File: rero_ils/modules/loans/transitions.py [hunk 11]
```diff
@@ -31,55 +31,89 @@
...
+    :param item_pid: a dict containing `value` and `type` fields to
+        uniquely identify the item.
+    """
     document_pid = get_document_pid_by_item_pid(item_pid)
-    # check if document has no other items
     document = Document.get_record_by_pid(document_pid)
     if document.get_number_of_items() == 1:
         for pending_loan in get_pending_loans_by_doc_pid(document_pid):
-            pending_loan['item_pid'] = item_pid
+            pending_loan["item_pid"] = item_pid
```
Comment (zannkukai) [inline rero_ils/modules/loans/transitions.py:43]: double quote is valid but we use single quote everywhere else

File: rero_ils/modules/loans/transitions.py [hunk 12]
```diff
@@ -31,55 +31,89 @@
...
-        super(ItemInTransitHouseToItemReturned, self).before(loan, **kwargs)
+        super().before(loan, **kwargs)
 
-        _ensure_same_location(loan['item_pid'],
-                              loan['transaction_location_pid'],
-                              self.dest,
-                              error_msg="Item should be in transit to house. ")
+        _ensure_same_location(
+            loan["item_pid"],
+            loan["transaction_location_pid"],
```
Comment (zannkukai) [inline rero_ils/modules/loans/transitions.py:72]: single quote

File: rero_ils/modules/loans/transitions.py [hunk 13]
```diff
@@ -31,55 +31,89 @@
...
-        super(ItemOnLoanToItemReturned, self).before(loan, **kwargs)
+        super().before(loan, **kwargs)
 
-        _ensure_same_location(loan['item_pid'],
-                              loan['transaction_location_pid'],
-                              self.dest,
-                              error_msg="Item should be in transit to house. ")
+        _ensure_same_location(
+            loan["item_pid"],
+            loan["transaction_location_pid"],
```
Comment (zannkukai) [inline rero_ils/modules/loans/transitions.py:107]: single quote

File: rero_ils/modules/loans/transitions.py [hunk 14]
```diff
@@ -31,55 +31,89 @@
...
+            self.dest,
+            error_msg="Item should be in transit to house.",
+        )
 
         # set end loan date as transaction date when completing loan
-        loan['end_date'] = loan['transaction_date']
+        loan["end_date"] = loan["transaction_date"]
```
Comment (zannkukai) [inline rero_ils/modules/loans/transitions.py:113]: single quote

File: rero_ils/modules/loans/transitions.py [hunk 15]
```diff
@@ -31,55 +31,89 @@
...
 
     def after(self, loan):
-        """Convert dates to string before saving loan."""
-        loan['end_date'] = loan['end_date'].isoformat()
-        super(ItemOnLoanToItemReturned, self).after(loan)
-        _update_document_pending_request_for_item(loan['item_pid'])
+        """Check for pending requests on this item after check-in."""
+        super().after(loan)
+        if self.assign_item:
+            _update_document_pending_request_for_item(loan["item_pid"])
```
Comment (zannkukai) [inline rero_ils/modules/loans/transitions.py:119]: same here

File: rero_ils/modules/loans/utils.py [hunk 16]
```diff
@@ -144,7 +145,43 @@ def is_item_available_for_checkout(item_pid):
...
+def loan_build_document_ref(loan_pid, loan):
+    """Build $ref for the Document attached to the Loan."""
+    base_url = current_app.config.get('RERO_ILS_APP_BASE_URL')
+    url_api = '{base_url}/api/{doc_type}/{pid}'
+    return {
+        '$ref': url_api.format(
+            base_url=base_url,
+            doc_type='documents',
+            pid=loan.document_pid)
+    }
```
Comment (zannkukai) [inline rero_ils/modules/loans/utils.py:187]: I created a general method to do this in `modules/utils.py:get_ref_for_pid`. See usage here : https://github.com/rero/rero-ils/blob/dev/tests/ui/test_utils_app.py

File: rero_ils/modules/patrons/api.py [hunk 17]
```diff
@@ -240,10 +240,10 @@ def patron_type_pid(self):
...
         """Get number of loans."""
-        search = current_circulation.loan_search_cls
-        search = search.filter("term", patron_pid=self.pid)
         exclude_states = ['CANCELLED', 'ITEM_RETURNED']
-        search = search.exclude("terms", state=exclude_states)
+        search_cls = current_circulation.loan_search_cls
+        search = search_cls().filter('term', patron_pid=self.pid)
+        search = search.exclude('terms', state=exclude_states)
         results = search.source().count()
         return results
```
Comment (zannkukai) [inline rero_ils/modules/patrons/api.py:248]: same as previous

File: rero_ils/modules/loans/listener.py [hunk 18]
```diff
@@ -34,14 +34,14 @@ def enrich_loan_data(sender, json=None, record=None, index=None,
     """
     if index == '-'.join(
             [current_circulation.loan_search_cls.Meta.index, doc_type]):
-        item = Item.get_record_by_pid(record.get('item_pid'))
+        item = Item.get_record_by_pid(record.get('item_pid').get('value'))
```
Comment (blankoworld) [inline rero_ils/modules/loans/listener.py:37]: What if item_pid is None? Is that possible?

File: rero_ils/modules/loans/api.py (hunk 6 above)
Comment (BadrAly) [reply rero_ils/modules/loans/api.py:115]: These two functions need to be here for the invenio-circulation to work, sorry. it is related to the fact that our class `Loan(IlsRecord)` and not `Loan(Loan)` from invenio-circulation. We may be able to change this soon.

**Pull request merged**